### ISSUE TITLE
Python Array API Compatibility Tracker

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#58743]
Progress tracker for all compatibility issues with [Python Array API specification]():

- [ ] Support for uint16, uint32, and uint64 (#58734)
- [ ] type promotion with 0d-tensors diverges from array API specification (#58736)
- [ ] Support for important numeric constants (#58739)
- [ ] support for scalars in logical_(and|not|or|xor) (#58740)
- [ ] torch.size() diverges from array API specification (#58741)
- [ ] Operator tracker for array API compatibility (#58742)

**Comment by leofang** [comment]
cc: @kmaehashi @emcastillo @asi1024

**Comment by pmeier** [comment]
> We've investigated each API in the Data API specification and summarized modifications needed to run code written using the API on PyTorch.
> Here is the full list.
https://docs.google.com/spreadsheets/d/13amsIeHhe38apCeZrOmsbsHT-aysmYur9-cugw5jVg0/edit#gid=0
In the list, we have classified each API into 6 categories:
> * No Changes: The current PyTorch implementation already conforms to the Data API.
> * New API: Need to add a new function or constant.
> * Enhancement: Need enhancement to the existing function (add new argument, etc.).
> * Enhancement with BC: Need to change signature of the existing function. (*breaking change* if we implement the API directly under `torch.*` namespace)
> * Behavior: Need to change behavior of the existing function. (*breaking change* if we implement the API directly under `torch.*` namespace)
> * N/A: The detailed specification is not determined yet in the Data API.
> 
> To implement the Data API directly under `torch.*` namespace, we need to discuss how to deal with the breaking changes. Major topics are:
> - Arguments: Some API cannot be compatible with the Data API specification without breaking changes. e.g., `torch.linspace(start, end, steps, ...)` v.s. Data API `linspace(start, stop, num, ...)`.
> - Argument names: PyTorch APIs conventionally use `dim`/`dims` as an argument name (e.g., `torch.stack`), whereas the Data API requires `axis` keyword argument.
> - Return value: For example, `torch.nonzero` returns a tensor when `as_tuple=False` (default) whereas the Data API requires to always return as tuple.
> - Behavior:
>   - `torch.cross` looks for the first dimension found with the size 3, whereas the Data API uses -1 as default axis.
>   - `floordiv` operator needs be changed to take floor. (The [documentation](https://pytorch.org/docs/stable/generated/torch.floor_divide.html) says `This behavior will be deprecated in a future PyTorch release` so it may be a good time to change the behavior)

_Originally posted by @asi1024 in https://github.com/pytorch/pytorch/issues/54581#issuecomment-810800837_

**Comment by asi1024** [comment]
@pmeier Sounds good to me!

**Comment by rgommers** [comment]
I have added a `module: python array api` label to gh-24422, so the bot will Cc people listed when that label gets added to new issues. Easier than manually Cc-ing a bunch of people. If you want to get added and don't have permissions to add yourself via editing that issue description, please let me know and I'll add you.

**Comment by asmeurer** [comment]
@rgommers can you add me?

**Comment by leofang** [comment]
+1

**Comment by kmaehashi** [comment]
@rgommers Could you add @asi1024 @emcastillo and me (@kmaehashi) to the list as well?

**Comment by asi1024** [comment]
Updated [the comparison table](https://docs.google.com/spreadsheets/d/13amsIeHhe38apCeZrOmsbsHT-aysmYur9-cugw5jVg0/edit?usp=sharing) to follow the current array API specification. Most of the interface details of linalg operations have been completed. Other major changes are as follows:
- additional features:
  * `broadcast_arrays(*arrays)`
  * `broadcast_to(x, /, shape)`
  * `can_cast(from_, to, /)`
  * `finfo(type, /)`: `smallest_normal`
- breaking change
  * The default value of `indexing` option of meshgrid

**Comment by rgommers** [comment]
Thanks for the update @asi1024!

> The default value of `indexing` option of meshgrid

This is gh-50276

**Comment by akern40** [comment]
Thank you to everyone who has put so much work towards this common Array API implementation in PyTorch! I have been trying to catch up with the discussion, and as best I can tell nobody has brought up the implementation of the [`__array_namespace__` function that is needed to generalize access](https://data-apis.org/array-api/latest/purpose_and_scope.html#how-to-adopt-this-api) to PyTorch's implementation of the Array API. Should this be added to the top-level to-do list? Or is it waiting until PyTorch is more compliant with the API? Please let me know if I'm missing something - thanks!

**Comment by rgommers** [comment]
Thanks for asking @akern40! `__array_namespace__` is listed in gh-58742. It's not exactly a new operator, so I'll move it to this issue. And you are right, it's the attribute that claims compliance, so it should be implemented when we reach (almost) full compliance with the standard.
